### ISSUE TITLE
Make magnetic bearing available for ride requests

### ISSUE THREAD

**Issue by kmadsen** [mapbox/mapbox-navigation-android#2901]
We are getting sensor events from both the [SensorEventEmitter](https://github.com/mapbox/mapbox-navigation-android/blob/40aa42ee161d8436e0dcc4ba1da3b23b716b8183/libnavigation-core/src/main/java/com/mapbox/navigation/core/sensors/SensorEventEmitter.kt#L29-L36) and the [LocationComponentCompassEngine](https://github.com/mapbox/mapbox-gl-native-android/blob/013076339111ae36862d474895588adec3795d24/MapboxGLAndroidSDK/src/main/java/com/mapbox/mapboxsdk/location/LocationComponentCompassEngine.java#L22).

Lets make a solution that provides magnetic bearing through the SDK. That way we can request directions with a better sense of direction.

This has been asked before, but I couldn't find the other github issues. It would be best if we could get this from NavNative. cc: @mskurydin @Guardiola31337 @LukasPaczos

**Comment by mskurydin** [comment]
@kmadsen could you elaborate a bit more on the design that involves nav-native?

I think, we only care about magnetic heading when we are stopped or moving extremely slowly. Detecting if we are stopped is something we can do inside nav-native. SDK can then ask nav-native about our last known bearing. This is something we had in mind before. If the hybrid router is inside nav-native, it would use this info internally, but would still require SDK to always pump magnetic heading into the native library.

cc @SiarheiFedartsou

**Comment by LukasPaczos** [comment]
The route request, and the options that its created from, is always initially passed from a user. The SDK only has control over re-route/faster route requests.

The SDK already uses the enhanced location update to execute re-routes, so as long as Nav Native is able to provide the `FixLocation`'s bearing value that's calculated from the sensor data (when appropriate), this use-case should already be handled. Just to double-check, is this already the case, or planned @mskurydin?

Similarly, our apps will handle this case automatically as well, as long as they build the initial route request using enhanced location updates.

**Comment by kmadsen** [comment]
> @ kmadsen could you elaborate a bit more on the design that involves nav-native?

Like lukas is saying, "The SDK already uses the enhanced location update to execute re-routes, so as long as Nav Native is able to provide the FixLocation's bearing value that's calculated from the sensor data (when appropriate), this use-case should already be handled"

The design is potentially handled, it'd be the guideline we'd want to formalize. Can we recommend using the enhanced location bearing to request directions? Or are we waiting for some projects to complete before we can do this?

"Detecting if we are stopped is something we can do inside nav-native." <== think this means we should wait for this project to complete? or is it complete?

**Comment by mskurydin** [comment]
If you query enhanced location while parked, you will not get good results (it will be taken from location provider as is and it's pretty chaotic on android).

The case with re-routes is different - we are **moving** when it happens and we have some history of location updates by this time. Both of these factors give us enough context to get a good bearing.

Fusing magnetometer into enhanced location is something that doesn't exist yet and this feature requires a careful design from the core side.

**Comment by LukasPaczos** [comment]
> Fusing magnetometer into enhanced location is something that doesn't exist yet and this feature requires a careful design from the core side.

Whenever this is available, the platform will automatically take advantage of it, so I don't think it's worth investing time in redoing the same functionality on the platform side right now.

@mskurydin @kmadsen if you agree, feel free to close the ticket.

**Comment by kmadsen** [comment]
Yeah this ticket wasn't meant to ask for something complex, but track a product being requested.

Sounds like it's not actionable by the sdk so good to close
